src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -429,74 +415,61 @@ private void basicAck(Channel channel, long deliveryTag) throws IOException {
+                                    "Sending via routing key '{}' is resumed. There are {}",
                                     routingKey,
-                                    exceededQueues.stream().map(QueueInfoWithVirtualLimit::getSizeDetails).collect(joining(", "))
+                                    queuesWithVirtualPublishLimit.getSizeDetails()
                             );
                         }
                     }
                 }
-            }
+            });
         } catch (Throwable t) {
             LOGGER.error("Error during check queue sizes", t);
         }
     }
 
-    private Map<String, List<QueueInfoWithVirtualLimit>> groupQueuesByRoutingKey(List<String> knownRoutingKeys) {
+    private Map<String, QueuesWithVirtualPublishLimit> groupQueuesByRoutingKey() {
         List<BindingInfo> bindings = new ArrayList<>();
-        knownExchanges.forEach(exchange -> bindings.addAll(
+        knownExchangesToRoutingKeys.forEach((exchange, routingKeys) -> bindings.addAll(
                 client.getBindingsBySource(rabbitMQConfiguration.getVHost(), exchange).stream()
-                        .filter(it -> it.getDestinationType() == QUEUE && knownRoutingKeys.contains(it.getRoutingKey()))
+                        .filter(it -> it.getDestinationType() == QUEUE && routingKeys.contains(it.getRoutingKey()))
                         .collect(Collectors.toList())
         ));
         Map<String, QueueInfo> queueNameToInfo = client.getQueues().stream()
                 .collect(toMap(QueueInfo::getName, Function.identity()));
-        Map<String, List<QueueInfoWithVirtualLimit>> routingKeyToQueues = new HashMap<>();
-        bindings.forEach(bindingInfo -> routingKeyToQueues
-                .computeIfAbsent(bindingInfo.getRoutingKey(), s -> new ArrayList<>())
-                .add(new QueueInfoWithVirtualLimit(
-                        queueNameToInfo.get(bindingInfo.getDestination()),
-                        queueNameToVirtualQueueLimit.get(bindingInfo.getDestination())
-                ))
-        );
-        return routingKeyToQueues;
+        Map<String, QueuesWithVirtualPublishLimit> routingKeyToQueuesWithLimit = new HashMap<>();
+        bindings.stream()
+                .collect(groupingBy(BindingInfo::getRoutingKey))
+                .forEach((routingKey, bindingsForRoutingKey) ->
+                        routingKeyToQueuesWithLimit.put(
+                                routingKey,
+                                new QueuesWithVirtualPublishLimit(
+                                        bindingsForRoutingKey.stream().map(bindingInfo -> queueNameToInfo.get(bindingInfo.getDestination())).collect(toList()),
+                                        connectionManagerConfiguration.getVirtualPublishLimit()
+                                )
+                        )
+                );
+        return routingKeyToQueuesWithLimit;
     }
```

Comment:
private Map<String, QueuesWithVirtualPublishLimit> groupQueuesByRoutingKey() {
        Map<String, QueueInfo> queueNameToInfo = client.getQueues().stream()
                .collect(toMap(QueueInfo::getName, Function.identity()));

        Map<String, List<BindingInfo>> bindings = knownExchangesToRoutingKeys.entrySet().stream()
                .flatMap(entry -> {
                            String exchange = entry.getKey();
                            Set<String> routingKeys = entry.getValue();
                            return client.getBindingsBySource(rabbitMQConfiguration.getVHost(), exchange).stream()
                                    .filter(it -> it.getDestinationType() == QUEUE && routingKeys.contains(it.getRoutingKey()));
                        }
                ).collect(groupingBy(BindingInfo::getRoutingKey));

        Map<String, QueuesWithVirtualPublishLimit> routingKeyToQueuesWithLimit = new HashMap<>();
        bindings.forEach((routingKey, bindingsForRoutingKey) ->
                        routingKeyToQueuesWithLimit.put(
                                routingKey,
                                new QueuesWithVirtualPublishLimit(
                                        bindingsForRoutingKey.stream().map(bindingInfo -> queueNameToInfo.get(bindingInfo.getDestination())).collect(toList()),
                                        connectionManagerConfiguration.getVirtualPublishLimit()
                                )
                        )
                );
        return routingKeyToQueuesWithLimit;
    }

Reply:
I think we can combine lockSendingIfSizeLimitExceeded and groupQueuesByRoutingKey into continuous stream or Kotlin sequence

Reply:
I didn't want to do it. I wanted to keep all logic with `com.rabbitmq.http.client.Client` in one place and do only locking in `lockSendingIfSizeLimitExceeded()`.
But I'll try if you want.

Reply:
I migrated `ConnectionManager` to Kotlin. Please see commit fbae68d71fe410ea5d427b10da82d5684895d85b.
But I don't think we can do it in one stream because we should take all queues for each routing key. I mean we can't do anything before check all queues.